launcher/settings/INIFile.cpp

```diff
@@ -183,3 +183,21 @@ void INIFile::set(QString key, QVariant val)
 {
     this->operator[](key) = val;
 }
+
+void INIFile::setList(QString key, QVariantList val)
+{
+    QString stringList = QJsonDocument(QVariant(val).toJsonArray()).toJson(QJsonDocument::Compact);
+
+    this->operator[](key) = stringList;
+}
+
+QVariantList INIFile::getList(QString key, QVariantList def) const
+{
+    if (this->contains(key)) {
+        auto src = this->operator[](key);
+
+        return QJsonDocument::fromJson(src.toByteArray()).toVariant().toList();
+    }
+
+    return def;
+}
```

Comment:
I played around with a LOT of ways to pull a list out of an INI value. This is one of the only ways the compiler would accept. everything else complained about `QVariant` not understanding the underlying type. Technically now that I understand how `QMetaType`'s work I could probably resolve this but frankly this seems clean in comparason.

Reply:
I think having separate functions for `QVariant` and `QVariantList` is confusing, especially considering `QVariantList` is technically accepted in `QVariant`.

Reply:
the problem is that there is no reall conversion between a `QStringList` and a `QVariant` or even a `QVariantList`
also the underlying `toString` that the INI file class uses to save things doens't understand how to save lists. The one time I got the type conversions to work I ended up with blank settings.

The JSON conversion is nessacery and this nessetates the split between `get` and `getList`

Reply:
After some tests, it seems that the only thing preventing `get` and `set` from working with those lists is the `INIFile` implementation thing. I think I said it before, but that thing is trash and I will stand by it until the dawn of time.
I have made a simple modification on it so that it's actually just a wrapper around a `QSettings` object, and it seems to work fine. I think it's a better alternative than adding a separate function and doing the JSON thing:

<details>
<summary>Le patch (note that I didn't remove the `getList` method, and neither changed the usages of it in `BaseInstance` and the test. Also, i didn't look at the `escape` and `unenscape` methods in INIFile, but maybe they can also be :hocho: )</summary>

```patch
diff --git a/launcher/minecraft/mod/tasks/LocalModParseTask.cpp b/launcher/minecraft/mod/tasks/LocalModParseTask.cpp
index da27a505..5342d693 100644
--- a/launcher/minecraft/mod/tasks/LocalModParseTask.cpp
+++ b/launcher/minecraft/mod/tasks/LocalModParseTask.cpp
@@ -242,7 +242,7 @@ ModDetails ReadQuiltModInfo(QByteArray contents)
     return details;
 }
 
-ModDetails ReadForgeInfo(QByteArray contents)
+ModDetails ReadForgeInfo(QString fileName)
 {
     ModDetails details;
     // Read the data
@@ -250,7 +250,7 @@ ModDetails ReadForgeInfo(QByteArray contents)
     details.mod_id = "Forge";
     details.homeurl = "http://www.minecraftforge.net/forum/";
     INIFile ini;
-    if (!ini.loadFile(contents))
+    if (!ini.loadFile(fileName))
         return details;
 
     QString major = ini.get("forge.major.number", "0").toString();
@@ -422,7 +422,7 @@ bool processZIP(Mod& mod, ProcessingLevel level)
             return false;
         }
 
-        details = ReadForgeInfo(file.readAll());
+        details = ReadForgeInfo(file.getFileName());
         file.close();
         zip.close();

diff --git a/launcher/settings/INIFile.cpp b/launcher/settings/INIFile.cpp
index e48e6f47..f7588fb8 100644
--- a/launcher/settings/INIFile.cpp
+++ b/launcher/settings/INIFile.cpp
@@ -42,6 +42,8 @@
 #include <QSaveFile>
 #include <QDebug>
 
+#include <QSettings>
+
 INIFile::INIFile()
 {
 }
@@ -99,75 +101,43 @@ QString INIFile::escape(QString orig)
 
 bool INIFile::saveFile(QString fileName)
 {
-    QByteArray outArray;
+    QSettings _settings_obj{ fileName, QSettings::Format::IniFormat };
+    _settings_obj.setFallbacksEnabled(false);
+
     for (Iterator iter = begin(); iter != end(); iter++)
-    {
-        QString value = iter.value().toString();
-        value = escape(value);
-        outArray.append(iter.key().toUtf8());
-        outArray.append('=');
-        outArray.append(value.toUtf8());
-        outArray.append('\n');
-    }
+        _settings_obj.setValue(iter.key(), iter.value());
+
+    _settings_obj.sync();
+
+    if (auto status = _settings_obj.status(); status != QSettings::Status::NoError) {
+        // Shouldn't be possible!
+        Q_ASSERT(status != QSettings::Status::FormatError);
+
+        if (status == QSettings::Status::AccessError)
+            qCritical() << "An access error occurred (e.g. trying to write to a read-only file).";
 
-    try
-    {
-        FS::write(fileName, outArray);
-    }
-    catch (const Exception &e)
-    {
-        qCritical() << e.what();
         return false;
     }
 
     return true;
 }
 
-
 bool INIFile::loadFile(QString fileName)
 {
-    QFile file(fileName);
-    if (!file.open(QIODevice::ReadOnly))
+    QSettings _settings_obj{ fileName, QSettings::Format::IniFormat };
+    _settings_obj.setFallbacksEnabled(false);
+
+    if (auto status = _settings_obj.status(); status != QSettings::Status::NoError) {
+        if (status == QSettings::Status::AccessError)
+            qCritical() << "An access error occurred (e.g. trying to write to a read-only file).";
+        if (status == QSettings::Status::FormatError)
+            qCritical() << "A format error occurred (e.g. loading a malformed INI file).";
         return false;
-    bool success = loadFile(file.readAll());
-    file.close();
-    return success;
-}
-
-bool INIFile::loadFile(QByteArray file)
-{
-    QTextStream in(file);
-#if QT_VERSION <= QT_VERSION_CHECK(6, 0, 0)
-    in.setCodec("UTF-8");
-#endif
-
-    QStringList lines = in.readAll().split('\n');
-    for (int i = 0; i < lines.count(); i++)
-    {
-        QString &lineRaw = lines[i];
-        // Ignore comments.
-        int commentIndex = 0;
-        QString line = lineRaw;
-        // Search for comments until no more escaped # are available
-        while((commentIndex = line.indexOf('#', commentIndex + 1)) != -1) {
-            if(commentIndex > 0 && line.at(commentIndex - 1) == '\\') {
-                continue;
-            }
-            line = line.left(lineRaw.indexOf('#')).trimmed();
-        }
-
-        int eqPos = line.indexOf('=');
-        if (eqPos == -1)
-            continue;
-        QString key = line.left(eqPos).trimmed();
-        QString valueStr = line.right(line.length() - eqPos - 1).trimmed();
-
-        valueStr = unescape(valueStr);
-
-        QVariant value(valueStr);
-        this->operator[](key) = value;
     }
 
+    for (auto&& key : _settings_obj.allKeys())
+        insert(key, _settings_obj.value(key));
+
     return true;
 }
 
diff --git a/launcher/settings/INIFile.h b/launcher/settings/INIFile.h
index 86bf0898..3745acd2 100644
--- a/launcher/settings/INIFile.h
+++ b/launcher/settings/INIFile.h
@@ -28,7 +28,6 @@ class INIFile : public QMap<QString, QVariant>
 public:
     explicit INIFile();
 
-    bool loadFile(QByteArray file);
     bool loadFile(QString fileName);
     bool saveFile(QString fileName);
```

</details>

Reply:
I'm happy to apply this patch but it still leaves me needing a template function to set and get `QList<T>` where `T` is not `QString`

there is no automatic conversion to `QVariant` for `QList<int>` for example

```cpp
template <typename T> QList<T>  INIFile::getList(QString key, QList<T> def) const
{   
    if (this->contains(key)) {
        QVariant src = this->operator[](key);
        QVariantList variantList = src.toList();

        QList<T> list_t;
        list_t.reserve(variantList.size());
        for (const QVariant& v : variantList)
        {
            list_t.append(v.value<T>());
        }
        return list_t;
    }

    return def;   
}

template <typename T> void INIFile::setList(QString key, QList<T> val)
{
    QVariantList variantList;
    variantList.reserve(val.size());
    for (const T& v : val)
    {
        variantList.append(v);
    }

    this->set(key, variantList);
}
```

Reply:
Patch applied, list functions moved to `QVariantHelpers`